monitor/sth_fetcher.go

```diff
@@ -75,8 +99,66 @@ type sthFetcher struct {
 
 	// How long to sleep between fetching the log's current STH
 	sthFetchInterval time.Duration
-	// How long to wait before giving up on an STH fetch
+	// How long to wait before giving up on an STH fetch or an STH consistency
+	// proof.
 	sthTimeout time.Duration
+
+	// prevSTH is the last STH that was fetched from the log
+	prevSTH *ct.SignedTreeHead
+
+	// sthLock is a Mutex for controlling updates to prevSTH
+	sthLock sync.Mutex
+
+	// verifier is used by verifySTHConsistency to prove consistency between two
+	// STHs
+	verifier sthFetcherVerifier
+}
+
+// newSTHFetcher returns an sthFetcher instance populated based on the provided
+// arguments
+func newSTHFetcher(
+	logger *log.Logger,
+	clk clock.Clock,
+	client monitorCTClient,
+	logURI string,
+	sthFetchInterval time.Duration,
+	sthTimeout time.Duration) *sthFetcher {
+	return &sthFetcher{
+		logger:           logger,
+		clk:              clk,
+		stats:            sthStats,
+		client:           client,
+		logURI:           logURI,
+		stopChannel:      make(chan bool),
+		sthFetchInterval: sthFetchInterval,
+		sthTimeout:       sthTimeout,
+		verifier:         merkle.NewLogVerifier(rfc6962.DefaultHasher),
+	}
+}
+
+// logErrorf formats a message to the sthFetcher's logger prefixed to identify
+// that an error occurred, that it was an sth-fetcher, and the logURI the error
+// affects
+func (f *sthFetcher) logErrorf(format string, args ...interface{}) {
+	// TODO(@cpu): We should be using os.Stderr here but doing so will mean
+	// changing integration tests. See
+	// https://github.com/letsencrypt/ct-woodpecker/issues/36
+	f.logger.Printf(
+		fmt.Sprintf("[ERROR] sth-fetcher %s : %s\n", f.logURI, format),
```

Comment:
There's a subtle bug here: If the logURI contains a Printf format character (which it just barely might, because of URL escaping), the outer Printf will give the wrong result.

I think it would be better to do the Sprintf first, assigning the result to a string, then do

    f.logger.Print("[ERROR]", "sth-fetcher", f.logURI, ":", line)

Reply:
That's a nice solution, thanks! Fixed in 43ce3d1